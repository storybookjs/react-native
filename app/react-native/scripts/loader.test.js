const path = require('path');
const { writeRequires, getMain, getPreviewExists } = require('./loader');

let pathMock;
let fileContentMock;

jest.mock('fs', () => ({
  ...jest.requireActual('fs'),
  writeFileSync: (filePath, fileContent, opts) => {
    pathMock = filePath;
    fileContentMock = fileContent;
  },
}));

jest.mock('prettier', () => ({
  format(s, opts) {
    return s;
  },
}));

describe('loader', () => {
  describe('getMain', () => {
    it('should return the main js default export as an object', () => {
      const main = getMain({ configPath: path.resolve(__dirname, 'mocks/all-config-files') });
      expect(main).toEqual({
        stories: ['./FakeStory.stories.tsx'],
        addons: [
          '@storybook/addon-ondevice-notes',
          '@storybook/addon-ondevice-controls',
          '@storybook/addon-ondevice-backgrounds',
          '@storybook/addon-ondevice-actions',
        ],
      });
    });

    it('should also work with relative paths', () => {
      // relative from where the command is run
      const main = getMain({ configPath: './scripts/mocks/all-config-files' });
      expect(main).toEqual({
        stories: ['./FakeStory.stories.tsx'],
        addons: [
          '@storybook/addon-ondevice-notes',
          '@storybook/addon-ondevice-controls',
          '@storybook/addon-ondevice-backgrounds',
          '@storybook/addon-ondevice-actions',
        ],
      });
    });
  });

  describe('getPreviewExists', () => {
    const supportedExtensions = ['js', 'jsx', 'ts', 'tsx'];
    describe('when using a relative path', () => {
      it('should return true if the preview exists', () => {
        supportedExtensions.forEach((ext) => {
          expect(getPreviewExists({ configPath: `scripts/mocks/preview-files/${ext}` })).toBe(true);
        });
      });

      it('should return false if the preview does not exist', () => {
        expect(getPreviewExists({ configPath: './scripts/mocks/no-preview' })).toBe(false);
      });

      it('should return false if the preview does not match any of supportedExtensions values', () => {
        expect(getPreviewExists({ configPath: './scripts/mocks/wrong-extension-preview' })).toBe(
          false
        );
      });
    });

    describe('when using an absolute path', () => {
      it('should return true if the preview exists', () => {
        supportedExtensions.forEach((ext) => {
          expect(
            getPreviewExists({
              configPath: path.resolve(__dirname, `mocks/preview-files/${ext}`),
            })
          ).toBe(true);
        });
      });

      it('should return false if the preview does not exist', () => {
        expect(getPreviewExists({ configPath: path.resolve(__dirname, 'mocks/no-preview') })).toBe(
          false
        );
      });

      it('should return false if the preview does not match any of supportedExtensions values', () => {
        expect(
          getPreviewExists({ configPath: path.resolve(__dirname, 'mocks/wrong-extension-preview') })
        ).toBe(false);
      });
    });
  });

  describe('writeRequires', () => {
    describe('when there is a story glob', () => {
      it('writes the story imports', () => {
        writeRequires({ configPath: 'scripts/mocks/all-config-files' });
        expect(pathMock).toEqual(
          path.resolve(__dirname, 'mocks/all-config-files/storybook.requires.js')
        );
        expect(fileContentMock).toMatchSnapshot();
      });
    });

    describe('when there is a story glob and exclude paths globs', () => {
      it('writes the story imports', () => {
        writeRequires({ configPath: 'scripts/mocks/exclude-config-files' });
        expect(pathMock).toEqual(
          path.resolve(__dirname, 'mocks/exclude-config-files/storybook.requires.js')
        );

        expect(fileContentMock).toContain('include-components/FakeStory.stories.tsx');
        expect(fileContentMock).not.toContain('exclude-components/FakeStory.stories.tsx');

        expect(fileContentMock).toMatchSnapshot();
      });
    });

    describe('when there is no story glob or addons', () => {
      it('writes no story imports or addons', () => {
        writeRequires({ configPath: 'scripts/mocks/blank-config' });
        expect(pathMock).toEqual(
          path.resolve(__dirname, 'mocks/blank-config/storybook.requires.js')
        );
        expect(fileContentMock).toMatchSnapshot();
      });
    });

    describe('when there is no preview', () => {
      it('does not add preview related stuff', () => {
        writeRequires({ configPath: 'scripts/mocks/no-preview' });
        expect(pathMock).toEqual(path.resolve(__dirname, 'mocks/no-preview/storybook.requires.js'));
        expect(fileContentMock).toMatchSnapshot();
      });
    });

    describe('when the absolute option is true', () => {
      it('should write absolute paths to the requires file', () => {
        writeRequires({ configPath: 'scripts/mocks/all-config-files', absolute: true });
        expect(pathMock).toEqual(
          path.resolve(__dirname, 'mocks/all-config-files/storybook.requires.js')
        );
        expect(fileContentMock).toContain(
          path.resolve(__dirname, 'mocks/all-config-files/FakeStory.stories.tsx')
        );
      });
    });
  });
});
