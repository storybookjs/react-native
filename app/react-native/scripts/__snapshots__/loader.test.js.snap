// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loader writeRequires when there are different file extensions writes the story imports 1`] = `
"
      /* do not change this file, it is auto generated by storybook. */

      import { configure, addDecorator, addParameters, addArgsEnhancer, clearDecorators } from '@storybook/react-native';

      global.STORIES = [{"titlePrefix":"","directory":"./scripts/mocks/file-extensions","files":"FakeStory.stories.tsx","importPathMatcher":"^\\\\.[\\\\\\\\/](?:scripts\\\\/mocks\\\\/file-extensions\\\\/FakeStory\\\\.stories\\\\.tsx)$"}]

      import "@storybook/addon-ondevice-notes/register";
import "@storybook/addon-ondevice-controls/register";
import "@storybook/addon-ondevice-backgrounds/register";
import "@storybook/addon-ondevice-actions/register";

      import { argsEnhancers } from "@storybook/addon-actions/dist/modern/preset/addArgs"

      
  import { decorators, parameters } from './preview';

  if (decorators) {
    if(__DEV__){
      // stops the warning from showing on every HMR 
      require('react-native').LogBox.ignoreLogs([
        '\`clearDecorators\` is deprecated and will be removed in Storybook 7.0',
      ]);
    }
    // workaround for global decorators getting infinitely applied on HMR, see https://github.com/storybookjs/react-native/issues/185
    clearDecorators();
    decorators.forEach((decorator) => addDecorator(decorator));
  }

  if (parameters) {
    addParameters(parameters);
  }


      
      try {
        argsEnhancers.forEach(enhancer => addArgsEnhancer(enhancer));
      } catch{}
    

      const getStories=() => {
        return {"./scripts/mocks/file-extensions/FakeStory.stories.tsx": require("./FakeStory.stories.tsx")};
      }

      configure(getStories, module, false)
  "
`;

exports[`loader writeRequires when there is a configuration object writes the story imports 1`] = `
"
      /* do not change this file, it is auto generated by storybook. */

      import { configure, addDecorator, addParameters, addArgsEnhancer, clearDecorators } from '@storybook/react-native';

      global.STORIES = [{"titlePrefix":"ComponentsPrefix","files":"**/*.stories.tsx","directory":"./scripts/mocks/configuration-objects/components","importPathMatcher":"^\\\\.[\\\\\\\\/](?:scripts\\\\/mocks\\\\/configuration-objects\\\\/components(?:\\\\/(?!\\\\.)(?:(?:(?!(?:^|\\\\/)\\\\.).)*?)\\\\/|\\\\/|$)(?!\\\\.)(?=.)[^/]*?\\\\.stories\\\\.tsx)$"}]

      import "@storybook/addon-ondevice-notes/register";
import "@storybook/addon-ondevice-controls/register";
import "@storybook/addon-ondevice-backgrounds/register";
import "@storybook/addon-ondevice-actions/register";

      import { argsEnhancers } from "@storybook/addon-actions/dist/modern/preset/addArgs"

      
  import { decorators, parameters } from './preview';

  if (decorators) {
    if(__DEV__){
      // stops the warning from showing on every HMR 
      require('react-native').LogBox.ignoreLogs([
        '\`clearDecorators\` is deprecated and will be removed in Storybook 7.0',
      ]);
    }
    // workaround for global decorators getting infinitely applied on HMR, see https://github.com/storybookjs/react-native/issues/185
    clearDecorators();
    decorators.forEach((decorator) => addDecorator(decorator));
  }

  if (parameters) {
    addParameters(parameters);
  }


      
      try {
        argsEnhancers.forEach(enhancer => addArgsEnhancer(enhancer));
      } catch{}
    

      const getStories=() => {
        return {"./scripts/mocks/configuration-objects/components/FakeStory.stories.tsx": require("./components/FakeStory.stories.tsx")};
      }

      configure(getStories, module, false)
  "
`;

exports[`loader writeRequires when there is a story glob and exclude paths globs writes the story imports 1`] = `
"
      /* do not change this file, it is auto generated by storybook. */

      import { configure, addDecorator, addParameters, addArgsEnhancer, clearDecorators } from '@storybook/react-native';

      global.STORIES = [{"titlePrefix":"","directory":"./scripts/mocks/exclude-config-files","files":"**/*.stories.tsx","importPathMatcher":"^\\\\.[\\\\\\\\/](?:scripts\\\\/mocks\\\\/exclude-config-files(?:\\\\/(?!\\\\.)(?:(?:(?!(?:^|\\\\/)\\\\.).)*?)\\\\/|\\\\/|$)(?!\\\\.)(?=.)[^/]*?\\\\.stories\\\\.tsx)$"}]

      import "@storybook/addon-ondevice-notes/register";
import "@storybook/addon-ondevice-controls/register";
import "@storybook/addon-ondevice-backgrounds/register";
import "@storybook/addon-ondevice-actions/register";

      import { argsEnhancers } from "@storybook/addon-actions/dist/modern/preset/addArgs"

      
  import { decorators, parameters } from './preview';

  if (decorators) {
    if(__DEV__){
      // stops the warning from showing on every HMR 
      require('react-native').LogBox.ignoreLogs([
        '\`clearDecorators\` is deprecated and will be removed in Storybook 7.0',
      ]);
    }
    // workaround for global decorators getting infinitely applied on HMR, see https://github.com/storybookjs/react-native/issues/185
    clearDecorators();
    decorators.forEach((decorator) => addDecorator(decorator));
  }

  if (parameters) {
    addParameters(parameters);
  }


      
      try {
        argsEnhancers.forEach(enhancer => addArgsEnhancer(enhancer));
      } catch{}
    

      const getStories=() => {
        return {"./scripts/mocks/exclude-config-files/include-components/FakeStory.stories.tsx": require("./include-components/FakeStory.stories.tsx")};
      }

      configure(getStories, module, false)
  "
`;

exports[`loader writeRequires when there is a story glob writes the story imports 1`] = `
"
      /* do not change this file, it is auto generated by storybook. */

      import { configure, addDecorator, addParameters, addArgsEnhancer, clearDecorators } from '@storybook/react-native';

      global.STORIES = [{"titlePrefix":"","directory":"./scripts/mocks/all-config-files","files":"FakeStory.stories.tsx","importPathMatcher":"^\\\\.[\\\\\\\\/](?:scripts\\\\/mocks\\\\/all-config-files\\\\/FakeStory\\\\.stories\\\\.tsx)$"}]

      import "@storybook/addon-ondevice-notes/register";
import "@storybook/addon-ondevice-controls/register";
import "@storybook/addon-ondevice-backgrounds/register";
import "@storybook/addon-ondevice-actions/register";

      import { argsEnhancers } from "@storybook/addon-actions/dist/modern/preset/addArgs"

      
  import { decorators, parameters } from './preview';

  if (decorators) {
    if(__DEV__){
      // stops the warning from showing on every HMR 
      require('react-native').LogBox.ignoreLogs([
        '\`clearDecorators\` is deprecated and will be removed in Storybook 7.0',
      ]);
    }
    // workaround for global decorators getting infinitely applied on HMR, see https://github.com/storybookjs/react-native/issues/185
    clearDecorators();
    decorators.forEach((decorator) => addDecorator(decorator));
  }

  if (parameters) {
    addParameters(parameters);
  }


      
      try {
        argsEnhancers.forEach(enhancer => addArgsEnhancer(enhancer));
      } catch{}
    

      const getStories=() => {
        return {"./scripts/mocks/all-config-files/FakeStory.stories.tsx": require("./FakeStory.stories.tsx")};
      }

      configure(getStories, module, false)
  "
`;

exports[`loader writeRequires when there is no preview does not add preview related stuff 1`] = `
"
      /* do not change this file, it is auto generated by storybook. */

      import { configure, addDecorator, addParameters, addArgsEnhancer, clearDecorators } from '@storybook/react-native';

      global.STORIES = [{"titlePrefix":"","directory":"./scripts/mocks/no-preview","files":"FakeStory.stories.tsx","importPathMatcher":"^\\\\.[\\\\\\\\/](?:scripts\\\\/mocks\\\\/no-preview\\\\/FakeStory\\\\.stories\\\\.tsx)$"}]

      import "@storybook/addon-ondevice-notes/register";
import "@storybook/addon-ondevice-controls/register";
import "@storybook/addon-ondevice-backgrounds/register";
import "@storybook/addon-ondevice-actions/register";

      import { argsEnhancers } from "@storybook/addon-actions/dist/modern/preset/addArgs"

      

      
      try {
        argsEnhancers.forEach(enhancer => addArgsEnhancer(enhancer));
      } catch{}
    

      const getStories=() => {
        return {"./scripts/mocks/no-preview/FakeStory.stories.tsx": require("./FakeStory.stories.tsx")};
      }

      configure(getStories, module, false)
  "
`;

exports[`loader writeRequires when there is no story glob or addons writes no story imports or addons 1`] = `
"
      /* do not change this file, it is auto generated by storybook. */

      import { configure, addDecorator, addParameters, addArgsEnhancer, clearDecorators } from '@storybook/react-native';

      global.STORIES = []

      

      

      

      

      const getStories=() => {
        return {};
      }

      configure(getStories, module, false)
  "
`;
